### PR TITLE
@turnkey/wallet-stamper

### PR DESCRIPTION
**Pull request by taylorjdawson**
## Summary & Motivation

This PR introduces a package `@turnkey/wallet-stamper`, which is a straightforward interface that allows users to sign messages and recover public keys using their wallets. The primary objective is to enable stamping of requests using the user's signature, thus allowing signing of activity requests with Solana and Ethereum wallets for turnkey activity requests.

Currently, this implementation supports Solana and EVM-type wallets, as the methods to obtain the public key differ slightly between the two. These changes align the SDK with recent backend features that facilitate signing turnkey activity requests with wallets.

## How I Tested These Changes

- Added a wallet stamper example to demonstrate the implementation.
- Added unit tests to ensure functionality and stability.

## Did you add a changeset?
Yup!

### CODE REVIEW

File: examples/with-wallet-stamper/src/components/add-wallet.tsx [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    //   console.log('decodedPublicKey', decodedPublicKey);
+
+    //   const encodedPublicKey = bs58.encode(
+    //     Buffer.from(decodedPublicKey, 'hex')
+    //   );
+    //   console.log('encodedPublicKey', encodedPublicKey);
+
+    //   const pubK = new PublicKey(Buffer.from(decodedPublicKey, 'hex'));
+    //   console.log('pubK', pubK.toString());
+    // }
```
Comment (robin-turnkey) [inline examples/with-wallet-stamper/src/components/add-wallet.tsx:40]: I'm guessing this test code needs to be removed ^^

File: examples/with-wallet-stamper/src/components/connect-wallet.solana.tsx [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  WalletConnectButton,
+} from '@solana/wallet-adapter-react-ui';
+
+export function ConnectWallet() {
+  const { connected, connect, select } = useWallet();
+  const { setVisible } = useWalletModal();
+  // const { buttonState, onConnect, onDisconnect, onSelectWallet } =
+  //   useWalletMultiButton({
+  //     onSelectWallet: setWalletModalConfig,
+  //   });
```
Comment (robin-turnkey) [inline examples/with-wallet-stamper/src/components/connect-wallet.solana.tsx:16]: Remove commented code

File: packages/http/src/__generated__/services/coordinator/public/v1/public_api.swagger.json [hunk 3]
```diff
@@ -3928,12 +3929,47 @@
...
+          "type": "boolean",
+          "description": "Disable email recovery for the sub-organization"
+        },
+        "disableEmailAuth": {
+          "type": "boolean",
+          "description": "Disable email auth for the sub-organization"
+        }
+      },
+      "required": ["subOrganizationName", "rootUsers", "rootQuorumThreshold"]
+    },
```
Comment (robin-turnkey) [inline packages/http/src/__generated__/services/coordinator/public/v1/public_api.swagger.json:3966]: We shouldn't be pulling this in until that endpoint is live. If you `make -C proto sync` from the `main` branch of mono, it should remove this.

**Review by robin-turnkey — COMMENTED** [review]
The core example code looks good. Main thing is there are lots of lingering blocks of commented out code that we should probably remove since this is public-facing. Also, the CreateSubOrgV6 endpoint isn't available in prod yet, so that generated code and anything that uses it shouldn't be included either.

File: packages/http/src/__generated__/services/coordinator/public/v1/public_api.swagger.json (hunk 3 above)
Comment (taylorjdawson) [reply packages/http/src/__generated__/services/coordinator/public/v1/public_api.swagger.json]: Done!

File: examples/with-wallet-stamper/src/components/turnkey-provider.tsx [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+  async function addWalletAuthenticator(email: Email) {
+    if (publicKey) {
+      const decodedPublicKey = Buffer.from(publicKey?.toBuffer()).toString(
+        "hex"
+      );
+      const res = await passkeyClient?.createApiKeys({
+        type: "ACTIVITY_TYPE_CREATE_API_KEYS_V2",
+        timestampMs: new Date().getTime().toString(),
+        organizationId: "f45c3014-e68c-40e2-a9a3-f4a36d5a0251",
```
Comment (robin-turnkey) [inline examples/with-wallet-stamper/src/components/turnkey-provider.tsx:110]: Hard-coded value; I think you need to thread the needle on this to get it to populate with local data

File: examples/with-wallet-stamper/src/components/turnkey-provider.tsx [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        },
+      });
+      console.log({ res });
+    }
+  }
+
+  async function signInWithWallet(email: Email) {
+    if (walletClient) {
+      const wallets = await walletClient?.getWallets({
+        organizationId: "f45c3014-e68c-40e2-a9a3-f4a36d5a0251",
```
Comment (robin-turnkey) [inline examples/with-wallet-stamper/src/components/turnkey-provider.tsx:130]: Hard-coded value

File: examples/with-wallet-stamper/src/components/turnkey-provider.tsx [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        parameters: {
+          apiKeys: [
+            {
+              apiKeyName: "wallet-authenticator",
+              publicKey: decodedPublicKey,
+              //@ts-ignore
+              curveType: "API_KEY_CURVE_ED25519",
+            },
+          ],
+          userId: "ac81ee4d-0a57-443b-a582-33cd2d7dd1ae",
```
Comment (robin-turnkey) [inline examples/with-wallet-stamper/src/components/turnkey-provider.tsx:120]: Hard-coded value

File: examples/with-wallet-stamper/src/components/turnkey-provider.tsx [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        type: "ACTIVITY_TYPE_CREATE_API_KEYS_V2",
+        timestampMs: new Date().getTime().toString(),
+        organizationId: "f45c3014-e68c-40e2-a9a3-f4a36d5a0251",
+        parameters: {
+          apiKeys: [
+            {
+              apiKeyName: "wallet-authenticator",
+              publicKey: decodedPublicKey,
+              //@ts-ignore
+              curveType: "API_KEY_CURVE_ED25519",
```
Comment (robin-turnkey) [inline examples/with-wallet-stamper/src/components/turnkey-provider.tsx:117]: I know we're just supporting Solana right now, so hard-coded curve value is fine for now, but can we store this in a constant somewhere to make it more clear to a client that this should be dynamically set?

File: examples/with-wallet-stamper/src/components/connect-wallet.tsx [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+    //   if (wallet?.adapter.name) {
+    //     console.log('connect sol');
+    //     select(wallet?.adapter.name);
+    //     await connectSolana();
+    //   }
+    // } else {
+    //   //
+    // }
+  };
```
Comment (robin-turnkey) [inline examples/with-wallet-stamper/src/components/connect-wallet.tsx:89]: 🧹

**Review by robin-turnkey — COMMENTED** [review]
The main flow is pretty close to working for me. Main thing is there are some hard-coded values that should be pointing the env vars, as well as a bit of commented code that should be cleaned up. I'd also suggest making the labels on the page a bit more straightforward and having the README give more direction on how the user should interact with the example. The relationship between the three forms on the screen was a bit unclear to me when I first looked at it.

The nuts and bolts look great though!

File: packages/wallet-stamper/src/constants.ts [hunk 9]
```diff
@@ -0,0 +1,7 @@
+export const SIGNATURE_SCHEME_TK_API_SECP256K1 =
+  "SIGNATURE_SCHEME_TK_API_SECP256K1";
+export const SIGNATURE_SCHEME_TK_API_ED25519 =
+  "SIGNATURE_SCHEME_TK_API_ED25519";
+export const WALLET_TYPE_EVM = "evm";
+export const WALLET_TYPE_SOLANA = "solana";
+export const STAMP_HEADER_NAME = "X-Stamp";
```
Comment (andrewkmin) [inline packages/wallet-stamper/src/constants.ts:7]: Doesn't have to be a now thing, but could be useful to have a separate header type here, such that we have the following:

```
X-Stamp: API keys
X-Stamp-Webauthn: webauthn/yubikey
X-Stamp-Wallet: crypto wallets
```

File: packages/wallet-stamper/src/types.ts [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+/**
+ * @interface TStamper
+ * @property {function(string): Promise<TStamp>} stamp - Function to stamp an input string.
+ */
+export interface TStamper {
+  stamp: (input: string) => Promise<TStamp>;
+}
+
+/**
+ * Base interface for wallet functionalities common across different blockchain technologies.
```
Comment (andrewkmin) [inline packages/wallet-stamper/src/types.ts:20]: Tiny nit: I think we can use slightly more casual language, like "chains" or "ecosystems" etc :)

File: packages/wallet-stamper/package.json [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    ".": {
+      "types": "./dist/index.d.ts",
+      "import": "./dist/index.mjs",
+      "require": "./dist/index.js",
+      "default": "./dist/index.mjs"
+    }
+  },
+  "types": "./dist/index.d.ts",
+  "license": "Apache-2.0",
+  "description": "API key stamper for @turnkey/http",
```
Comment (andrewkmin) [inline packages/wallet-stamper/package.json:16]: Tiny nit: "Wallet stamper"

File: packages/wallet-stamper/README.md [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+3. **Instantiate the WalletStamper**: Create an instance of the `WalletStamper` using the wallet interface.
+
+4. **Instantiate the TurnkeyClient**: Create an instance of the `TurnkeyClient` with the `WalletStamper` instance.
+
+5. **Stamp Requests**: Now when making request using the TurnkeyClient, the wallet stamper will automatically stamp the request with the user's wallet public key and signature.
+
+### Example: Signing with a Solana Wallet
+
+In this example, we are using a local Solana wallet.
+For information on using an injected Solana wallet such as Solflare, please refer to the [`with-wallet-stamper`](../../examples/with-wallet-stamper) example.
```
Comment (andrewkmin) [inline packages/wallet-stamper/README.md:89]: Just curious -- any particular reason for solflare over others like phantom/backpack?

File: packages/wallet-stamper/.env.example [hunk 13]
```diff
@@ -0,0 +1,5 @@
+# Public API key, obtain from your API provider
```
Comment (andrewkmin) [inline packages/wallet-stamper/.env.example:1]: This env file is only needed for local testing correct?

File: examples/with-wallet-stamper/src/lib/types.ts [hunk 14]
```diff
@@ -0,0 +1,15 @@
+import { type TurnkeyClient } from "@turnkey/http";
+
+export type Attestation = Parameters<
+  TurnkeyClient["createSubOrganization"]
+>[0]["parameters"]["rootUsers"][0]["authenticators"][0]["attestation"];
+
+export interface PassKeyRegistrationResult {
```
Comment (andrewkmin) [inline examples/with-wallet-stamper/src/lib/types.ts:7]: Teeeeniest tiniest of nits: `Passkey` for consistency

File: packages/wallet-stamper/src/__tests__/index.test.ts
Comment (andrewkmin) [inline packages/wallet-stamper/src/__tests__/index.test.ts:1]: Appreciate the test coverage as well!!

File: packages/wallet-stamper/src/__tests__/constants.ts [hunk 15]
```diff
@@ -0,0 +1,16 @@
+// Constants for Ethereum keys
+export const ETHEREUM_PUBLIC_KEY =
+  "0x048318535b54105d4a7aae60c08fc45f9687181b4fdfc625bd1a753fa7397fed753547f11ca8696646f2f3acb08e31016afac23e630c5d11f59f61fef57b0d2aa5";
+export const ETHEREUM_PRIVATE_KEY =
+  "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80";
```
[the remark this answers is not in the document]
Comment (taylorjdawson) [reply packages/wallet-stamper/src/__tests__/constants.ts:5]: haha its a test account but good call 😄

File: packages/wallet-stamper/src/constants.ts (hunk 9 above)
Comment (taylorjdawson) [reply packages/wallet-stamper/src/constants.ts:7]: ooo yes I like that idea cc @robin-turnkey

File: packages/wallet-stamper/README.md (hunk 12 above)
Comment (taylorjdawson) [reply packages/wallet-stamper/README.md:89]: no actually just the one that I reached for when testing!

Comment (taylorjdawson) [reply packages/wallet-stamper/README.md:89]: should we list more?

File: packages/wallet-stamper/.env.example (hunk 13 above)
Comment (taylorjdawson) [reply packages/wallet-stamper/.env.example:1]: yup! Ill add comment

File: examples/with-wallet-stamper/src/components/connect-wallet.tsx (hunk 8 above)
Comment (robin-turnkey) [inline examples/with-wallet-stamper/src/components/connect-wallet.tsx:89]: Is this comment being left here intentionally? If so, can you document the purpose of it more clearly?

**Review by robin-turnkey — APPROVED** [review]
Approved with small, cosmetic suggestions

File: examples/with-wallet-stamper/README.md
[the remark this answers is not in the document]
Comment (taylorjdawson) [reply examples/with-wallet-stamper/README.md:1]: good call!

**Review by robin-turnkey — APPROVED** [review]

File: examples/with-wallet-stamper/package.json [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+    "@solana/wallet-adapter-base": "^0.9.23",
+    "@solana/wallet-adapter-base-ui": "^0.1.2",
+    "@solana/wallet-adapter-react": "^0.15.35",
+    "@solana/wallet-adapter-react-ui": "^0.9.35",
+    "@solana/wallet-adapter-wallets": "^0.19.32",
+    "@solana/web3.js": "^1.88.1",
+    "@t3-oss/env-nextjs": "^0.11.0",
+    "@turnkey/api-key-stamper": "workspace:*",
+    "@turnkey/encoding": "workspace:^",
+    "@turnkey/http": "^2.10.0",
```
Comment (andrewkmin) [inline examples/with-wallet-stamper/package.json:32]: Do we want to use `workspace` for these other Turnkey deps as well?

File: packages/wallet-stamper/src/types.ts [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+/**
+ * EVM wallets require a signed message to derive the public key.
+ * @interface EvmWalletInterface
+ * @extends BaseWalletInterface
+ * @property {function(string, string): Promise<string>} recoverPublicKey - Public key recovery from signature and message, returning a hex encoded SECP256K1 public key as a string.
+ * @property {'evm'} type - The type of the wallet.
+ */
+export interface EvmWalletInterface extends BaseWalletInterface {
+  recoverPublicKey: (message: string, signature: string) => Promise<string>;
+  type: "evm";
```
Comment (andrewkmin) [inline packages/wallet-stamper/src/types.ts:49]: tiny nit: can use/import `WALLET_TYPE_EVM`

**Review by andrewkmin — APPROVED** [review]
Looks great overall 🔥🔥🔥

Just some minor things to be aware of:
- Potential `X-Stamp` change as followup
- Versions in example apps (should we start pinning versions?)
- Generated files: should we update these based on today's release? (can be in a separate PR)

File: examples/with-wallet-stamper/src/lib/ethereum.ts [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+  }
+
+  const signature = await client.signMessage({
+    account,
+    message,
+  });
+  return signature;
+};
+
+export const recoverPublicKey = () => {};
```
[the remark this answers is not in the document]
Comment (taylorjdawson) [reply examples/with-wallet-stamper/src/lib/ethereum.ts:32]: Actually I'll remove this for now as its not being used!

**Pull request merged**